### PR TITLE
enhance consolidate training run outputs into a single runs/ directory

### LINKED ISSUE #293: Consolidate Training/Evaluation Run Outputs into a Single `runs/` Directory
## Problem

A single training run scatters output across five unrelated directories:

| Artifact | Current location |
|---|---|
| Model checkpoints | `saved_models/<run-name>/` |
| Lightning CSV logs | `lightning_logs/version_N/` (auto-incremented, not named) |
| W&B offline cache | `wandb/offline-run-<timestamp>-<id>/` |
| Eval artifacts (PDFs, CSVs, `.pt` files) | `wandb/` or `mlruns/` via `self.logger.save_dir` |
| MLFlow artifacts | `mlruns/` (hard-coded in `CustomMLFlowLogger.save_dir`) |

## Run Name Format

Auto-generated in `neural_lam/train_model.py` as:
`{prefix}{model}-{processor_layers}x{hidden_dim}-{MM_DD_HH}-{random_4digits}`
e.g. `train-graph_lam-2x64-02_21_12-4571`. Override with `--logger_run_name`.
The run name is applied to `saved_models/` and the external logger, but **not** to `lightning_logs/`.

## Proposed Change

All output should go under `runs/<run-name>/`:

1. **`train_model.py`**: set `ModelCheckpoint(dirpath=f"runs/{run_name}/checkpoints")` and `default_root_dir=f"runs/{run_name}"` on `pl.Trainer`.
2. **`utils.py`**: pass `save_dir=f"runs/{run_name}"` to `WandbLogger` and `CustomMLFlowLogger`.
3. **`custom_loggers.py`**: make `save_dir` return the run-scoped path instead of `"mlruns"`.
4. **`ar_model.py`**: no changes needed — artifacts already use `self.logger.save_dir`.
5. **`README.md`** / **`.gitignore`**: replace `saved_models/`, `lightning_logs/`, `wandb/` references with `runs/`.

### PR DESCRIPTION
**Pull request by sudhansu-24**
## Describe your changes

Training and evaluation artifacts are written under a single directory `runs/<run-name>/`: `ModelCheckpoint` uses `runs/<run-name>/checkpoints/`, `Trainer(default_root_dir=...)` keeps Lightning CSV logs under that run instead of a top-level `lightning_logs/`, and `WandbLogger` / `CustomMLFlowLogger` use `save_dir=run_dir` so internal logger paths and code using `self.logger.save_dir` (e.g. plots) stay under the run root. Checkpoints remain outside W&B’s `wandb/` subtree so large files are not synced by default.

**Motivation:** Issue #293 and maintainer feedback (W&B selective sync, common run root, MLflow temp images not in CWD).
**Dependencies:** None

## Issue Link

closes #293 

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [x] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by joeloskarsson** [comment]
@sadamov assigning you here to decide later if this should be closed in favor of #297 or what is the best path forward with this.

**Comment by sudhansu-24** [comment]
thanks @sadamov for clarifying.

I’ll continue implementation/revisions on #580 and coordinate here.
@Shyam-Sunder-saini @techaadii, if you have any pending changes or preferences from your earlier work that should be included please share them and i will incorporate them into this pr so we can converge quickly

**Comment by Shyam-Sunder-saini** [comment]
Thanks @sudhansu-24  for taking this forward!

From my side, I’ve aligned all training artifacts so they are now scoped under `runs/<run-name>/`, including checkpoints, Lightning logs, W&B, and MLflow outputs.

I also updated the logger setup so that both WandbLogger and CustomMLFlowLogger use the same `save_dir` (run directory). Additionally, MLflow now falls back to the run directory if `MLFLOW_TRACKING_URI` is not set.

Currently, only one logger is active at a time (default is W&B), and MLflow artifacts are generated when explicitly running with `--logger mlflow`.

If there are any preferences around structure or logging behavior from earlier work, I’m happy to incorporate them.

Let me know if you’d like me to push any additional changes to the PR.

**Comment by sudhansu-24** [comment]
Thanks @Shyam-Sunder-saini   

could you share the exact changes you want added beyond the current #580 state (especially around the `MLFLOW_TRACKING_URI` fallback) either as:

a short checklist by file, or a commit/PR branch we can cherry-pick from?

if you post that i will incorporate it quickly so we can finalize review.

**Comment by sudhansu-24** [comment]
Hi @sadamov noticed the duplicate label and i am a bit confused 

#580 was opened on Apr 4, and as per your coordination note on #293 I asked @Shyam-Sunder-saini and @techaadii to contribute their changes here so we could converge, now i found that there's been seperate PR #586 opened on Apr 9 by @Shyam-Sunder-saini despite me asking any changes they want on top of my PR so i could incoporate it in #580 itself.

If you'd prefer to land via #586, let me know and I'll close this. Otherwise I'll wait for your review.

**Comment by sadamov** [comment]
@sudhansu-24 i fixed the labels, could you also fix the title? This is not a bug but an enhancement. you are no both assigned and we will continue with the implementation here in this PR.

**Comment by sudhansu-24** [comment]
@sadamov Thanks for clarification!
Updated the PR title and will continue on #580 and address any remaining review feedback here.

**Review by sadamov — COMMENTED** [review]
Implementation is clean and minimal. Inline suggestions below. A few points can't anchor inline because the touched lines are outside the diff hunks, so noting them here:

- README.md wasn't touched and still references the old layout (#293 scoped this). Line 407 says `wandb/dryrun...` (no longer produced); line 440 says checkpoints live in `saved_models/` (now under `runs/<run-name>/checkpoints/`). Worth a small README pass in this PR.
- `setup_training_logger` in `neural_lam/utils.py` gained a `run_dir` parameter but the docstring Parameters section wasn't updated. Please add a `run_dir : str` entry describing it as the directory under which all run artifacts are written.
- `tests/test_cli.py::test_wandb_logger_kwargs` now passes `run_dir` but doesn't assert it reaches `WandbLogger`. One extra `assert kwargs["save_dir"] == "runs/my-run"` next to the existing kwargs assertions closes the loop.

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,6 @@
 ### Project Specific ###
-wandb
-saved_models
-lightning_logs
+runs/
+lightning_logs/
```
Comment (sadamov) [inline .gitignore]: Every existing user has `saved_models/`, `wandb/`, and possibly `lightning_logs/` directories on disk from pre-PR runs. Dropping the ignores means their first `git status` after pulling surfaces a wall of untracked files. Suggest keeping the legacy entries — cost is two lines, benefit is zero churn for users.

```suggestion
### Project Specific ###
runs/
saved_models
lightning_logs
wandb
```

File: neural_lam/custom_loggers.py [hunk 2]
```diff
@@ -15,10 +16,11 @@ class CustomMLFlowLogger(pl.loggers.MLFlowLogger):
     of version `2.0.3` at least.
     """
 
-    def __init__(self, experiment_name, tracking_uri, run_name):
+    def __init__(self, experiment_name, tracking_uri, run_name, save_dir=None):
         super().__init__(
             experiment_name=experiment_name, tracking_uri=tracking_uri
         )
+        self._save_dir = save_dir or "mlruns"
```
Comment (sadamov) [inline neural_lam/custom_loggers.py]: The `or "mlruns"` fallback silently recreates an `mlruns/` directory at CWD if a caller forgets `save_dir` — exactly the inconsistency #293 aims to remove. Prefer making it required so misuse fails loudly.

```suggestion
    def __init__(self, experiment_name, tracking_uri, run_name, save_dir):
        super().__init__(
            experiment_name=experiment_name, tracking_uri=tracking_uri
        )
        self._save_dir = save_dir
```

File: neural_lam/custom_loggers.py [hunk 3]
```diff
@@ -57,7 +59,8 @@ def log_image(self, key, images, step=None):
 
         # Need to save the image to a temporary file, then log that file
         # mlflow.log_image, should do this automatically, but is buggy
-        temporary_image = f"{key}.png"
+        os.makedirs(self.save_dir, exist_ok=True)
+        temporary_image = os.path.join(self.save_dir, f"{key}.png")
```
Comment (sadamov) [inline neural_lam/custom_loggers.py]: `os.makedirs` runs on every `log_image` call. Once per logger instance is enough — move it into `__init__` after `self._save_dir = save_dir` and drop it here.

```suggestion
        # Need to save the image to a temporary file, then log that file
        # mlflow.log_image, should do this automatically, but is buggy
        temporary_image = os.path.join(self.save_dir, f"{key}.png")
        images[0].savefig(temporary_image)
```

File: neural_lam/train_model.py [hunk 4]
```diff
@@ -338,12 +339,14 @@ def main(input_args=None):
             f"{time.strftime('%m_%d_%H')}-{random_run_id:04d}"
         )
 
+    run_dir = os.path.join("runs", run_name)
```
Comment (sadamov) [inline neural_lam/train_model.py]: Minor, non-blocking: `"runs"` is the only place the root is defined. A module-level `RUNS_ROOT = "runs"` makes future changes (or an optional CLI override) trivial.

```suggestion
    run_dir = os.path.join(RUNS_ROOT, run_name)
```

**Comment by sadamov** [comment]
Follow-up on the README gap mentioned in the review: the SLURM example at lines 494-495 writes to `lightning_logs/`  we should also adjust that to the new run folder:

```
#SBATCH --output=runs/slurm_logs/neurallam_out_%j.log
#SBATCH --error=runs/slurm_logs/neurallam_err_%j.log
```

**Comment by sudhansu-24** [comment]
thanks @sadamov  

addressed all four suggestions and the other three changes from review.
local `tests/test_cli.py` (7/7) and `tests/test_training.py::test_training[dummydata]` pass.

**Comment by sadamov** [comment]
@sudhansu-24  please fix the conflict then I can do the final review.

**Comment by sudhansu-24** [comment]
> @sudhansu-24 please fix the conflict then I can do the final review.

@sadamov done, ready for review.

**Review by sadamov — APPROVED** [review]
LGTM!

Ran `pytest` and `pre-commit` locally — there was a flake8 `E305` from the `RUNS_ROOT` constant being placed between two function definitions. Pushed a follow-up commit that:

- drops `RUNS_ROOT` and replaces it with a `--runs_root` CLI flag (default `"runs"`), so users who can't write to the repo root can override it;
- documents the eager `os.makedirs` side effect in `CustomMLFlowLogger.__init__`;
- mirrors the existing wandb-branch `save_dir` assertion in `test_wandb_id_ignored_with_mlflow_warns`.

**Review by sadamov — APPROVED** [review]
Marking this one as ready

File: neural_lam/utils.py [hunk 5]
```diff
@@ -552,6 +559,7 @@ def setup_training_logger(datastore: Any, args: Any, run_name: str) -> Any:
             config=dict(training=vars(args), datastore=datastore._config),
             resume=wandb_resume,
             id=args.wandb_id,
+            save_dir=run_dir,
         )
```
Comment (joeloskarsson) [inline neural_lam/utils.py:568]: What I am confused about here is that if we now set the wandb logger save_dir to `run_dir`, doesn't this mean that everything there gets synced to wandb? Including all checkpoints (was not the case before, when that was separate dir)? Seems like potentially uploading a lot of data.

Comment (sadamov) [reply neural_lam/utils.py:568]: WandbLogger checkpoint upload is gated on `log_model` (default `False`, not overridden here), independent of `save_dir`. And wandb only auto-syncs `<save_dir>/wandb/<run_id>/files/`, not the whole save_dir. Our `run_dir/checkpoints/` is a sibling of `run_dir/wandb/`, never inside it. So this just moves the local cache, cloud footprint unchanged.

Comment (joeloskarsson) [reply neural_lam/utils.py:568]: Aha, I misunderstood the exact directory structure then 😁 That seems perfect, thanks for explaining!

**Comment by sadamov** [comment]
@sudhansu-24 congrats, your PR just got merged! thanks for your work and your patience :heart:

**Comment by sudhansu-24** [comment]
> @sudhansu-24 congrats, your PR just got merged! thanks for your work and your patience ❤️

thankyou❤️
